### PR TITLE
expression: implement vectorized evalutaion for `builtinTiDBDecodeKeySig` 

### PR DESCRIPTION
**Pull request by TennyZhuang**
PCP #12106 

### What problem does this PR solve? 

#12106 

### What is changed and how it works?

```
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinInfoFunc/builtinTiDBDecodeKeySig-VecBuiltinFunc-8         	     681	   1502962 ns/op	  329383 B/op	    4884 allocs/op
BenchmarkVectorizedBuiltinInfoFunc/builtinTiDBDecodeKeySig-NonVecBuiltinFunc-8      	     579	   2085128 ns/op	  329384 B/op	    4884 allocs/op
PASS
ok  	github.com/pingcap/tidb/expression	2.666s

```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

Release note

 - vectorize `builtinTiDBDecodeKeySig`

### CODE REVIEW

File: expression/builtin_info_vec.go [hunk 1]
```diff
@@ -179,9 +180,28 @@ func (b *builtinVersionSig) vecEvalString(input *chunk.Chunk, result *chunk.Colu
...
+	buf, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	result.ReserveString(n)
+
```
Comment (wshwsh12) [inline expression/builtin_info_vec.go]: Useless line

File: expression/builtin_info_vec.go [hunk 2]
```diff
@@ -179,9 +180,28 @@ func (b *builtinVersionSig) vecEvalString(input *chunk.Chunk, result *chunk.Colu
...
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		result.AppendString(decodeKey(b.ctx, buf.GetString(i)))
+	}
+
```
Comment (wshwsh12) [inline expression/builtin_info_vec.go]: Useless line.

File: expression/builtin_info_vec_test.go [hunk 3]
```diff
@@ -33,10 +33,12 @@ var vecBuiltinInfoCases = map[string][]vecExprBenchCase{
 	ast.Database: {
 		{retEvalType: types.ETString, childrenTypes: []types.EvalType{}},
 	},
-	ast.User:          {},
-	ast.TiDBDecodeKey: {},
-	ast.RowCount:      {},
-	ast.CurrentRole:   {},
+	ast.User: {},
+	ast.TiDBDecodeKey: {
+		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETString}},
```
Comment (SunRunAway) [inline expression/builtin_info_vec_test.go]: Could you add a new test and implement a new `gener` reflecting the specific input string of `builtinTiDBDecodeKeySig`

**Comment by TennyZhuang** [comment]
@SunRunAway 

I have added a `tidbKeyGener`, and I'm sure all all decodeKey result is either index or table now (by debug).

```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinInfoFunc/builtinTiDBDecodeKeySig-VecBuiltinFunc-4         	     332	   3549957 ns/op	  888283 B/op	   13748 allocs/op
BenchmarkVectorizedBuiltinInfoFunc/builtinTiDBDecodeKeySig-NonVecBuiltinFunc-4      	     259	   4571547 ns/op	  888278 B/op	   13748 allocs/op
PASS
```

New benchmark

**Comment by TennyZhuang** [comment]
@SunRunAway Is the unit test failed caused by my PR?

**Comment by TennyZhuang** [comment]
Hi @SunRunAway 

It seems that the unit test failed due to some other error and will be fixed in https://github.com/pingcap/tidb/pull/13275

Can this PR be merged firstly?

**Pull request merged**